tests/unittest.py

```diff
@@ -106,11 +109,11 @@ def method_name(orig, *args, **kwargs):
         return orig(*args, **kwargs)
     """
 
-    def _around(code):
+    def _around(code: Callable[P, R]) -> None:
         name = code.__name__
         orig = getattr(target, name)
 
-        def new(*args, **kwargs):
+        def new(*args: P.args, **kwargs: P.kwargs) -> R:
             return code(orig, *args, **kwargs)
 
         setattr(target, name, new)
```

Comment:
A minor nitpick: Is this strictly correct? From the perspective of type annotations only:

- `code` is the wrapper function which is being decorated by `@around(target)`.
- The first argument to `code` is `orig`.
- Therefore the first argument of `new` is also `orig`.
- Therefore the call to `code(...)` on line +117 passes in `orig` twice. But that's not what the source code does.

I think `code: Callable[Concatenate[object, P], R]` might be more accurate (where `object` represents `orig`).

Reply:
That might be more accurate, yes. I'll check that!

Reply:
> Therefore the first argument of `new` is also `orig`.

I don't think this is correct -- the code would fail in this case because `setUp` doesn't accept `*args`.

`new` needs to match the signature of `orig`, not of `code`. I believe this is what the current code does.

Reply:
I think we're agreeing with each other, but I didn't express myself well.  Perhaps instead of

> Therefore the first argument of `new` is also `orig`.

I should have said

> *mypy can deduce that* the first argument of `new` has type `type(orig)`.

----

Putting that aside for the moment, however, I entirely agree with your summary here:

> `new` needs to match the signature of `orig`, not of `code`

but I don't think this is what the current set of annotations mean. At present, both `new` and `code` are of type `Callable[P, R]`. But this doesn't make sense, because `code` takes `orig` as an extra argument! 

I still think that `code: Callable[Concatenate[object, P], R]` is correct here. Have I managed to convince you?

(You could try writing/casting `orig: Callable[P, R] = ...`; I think mypy would detect the inconsistency I've tried to describe above.)

Reply:
I believe 603d70acd1fc1c5e479dff59e41d7fa73beea041 fixes this?